src/include/common/concurrent_pointer_vector.h

```diff
@@ -0,0 +1,280 @@
+     * @return copy of the iterator equal to this before increment
+     */
+    Iterator operator++(int) {
+      Iterator copy = *this;
+      operator++();
+      return copy;
+    }
+
+    /**
+     * Equality check.
+     * @param other other iterator to compare to
+     * @return if the two iterators point to the same spot
+     */
+    bool operator==(const Iterator &other) const {
+      if (LIKELY(other.is_end_)) {
+        return current_index_ >= end_index_;
+      }
+      if (LIKELY(is_end_)) {
+        return other.current_index_ >= other.end_index_;
+      }
+      TERRIER_ASSERT(vector_ == other.vector_, "should only compare iterators for the same vector");
+      return current_index_ == other.current_index_;
+    }
+    /**
+     * Equality check.
+     * @param other other iterator to compare to
+     * @return if the two iterators point to different spot
+     */
+    bool operator!=(const Iterator &other) const { return !operator==(other); }
+
+    /**
+     * Iterator default constructor, initializes to end of the vector
+     */
+    Iterator() = default;
+
+    /**
+     * Iterator constructor for iterator of certain vector, initializes to start of vector, and iterates to length
+     * of vector at time of creation
+     */
+    Iterator(const ConcurrentPointerVector *vector)  // NOLINT
+        : end_index_(vector->claimable_index_),
+          vector_(const_cast<ConcurrentPointerVector *>(vector)),
+          is_end_(false) {}
+    ~Iterator() = default;
+
+   private:
+    uint64_t current_index_ = 0, end_index_ = 0;
+    ConcurrentPointerVector *vector_ = nullptr;
+    T **local_value_ = nullptr;
+    bool is_end_ = true;
+  };
+
+  /**
+   * begin
+   * @return Iterator at the begining of vector
+   */
+  Iterator begin() const { return {this}; }  // NOLINT
+
+  /**
+   * @return Iterator at the end of vector
+   */
+  Iterator end() const { return END; }  // NOLINT
+
+ protected:
+  /**
+   * RESIZE_FACTOR factor of resize when capacity is reached in vector
+   */
+  static constexpr uint64_t RESIZE_FACTOR = 2;
+
+  /**
+   * START_SIZE default initial size of vector
+   */
+  static constexpr uint64_t START_SIZE = 8;
+
+ private:
+  static constexpr Iterator END = {};
+  static constexpr uint64_t SHIFT_AMOUNT = 63;
+  static constexpr uint64_t READABLE_FLAG = static_cast<uint64_t>(1) << SHIFT_AMOUNT;
+  // the first bit in the pointer is used as a flag to denote whether an item has been inserted into this
+  void SetReadable(T **array, uint64_t i) {
+    array[i] = reinterpret_cast<T *>(reinterpret_cast<uint64_t>(array[i]) | READABLE_FLAG);
+  }
+
+  void SetNotReadable(T **array, uint64_t i) {
+    array[i] = reinterpret_cast<T *>(reinterpret_cast<uint64_t>(array[i]) & ~static_cast<uint64_t>(READABLE_FLAG));
+  }
+
+  bool GetReadability(T **array, uint64_t i) const {
+    return static_cast<bool>(reinterpret_cast<uint64_t>(array[i]) >> SHIFT_AMOUNT);
+  }
+
+  std::atomic<uint64_t> capacity_ = 0;
```

Comment:
Could you add a comment saying that this is protected by resize_mutex_?

Reply:
done